### DISCUSSION TITLE
How to Parallelize Agents in a GroupChat Workflow

### DISCUSSION THREAD

**Discussion by anusonawane** [microsoft/autogen#4215] [Q&A] [answered]
I'm working on a project where I use the autogen library to coordinate multiple agents in a GroupChat workflow. The workflow currently involves:

- A Data Fetcher Agent (retrieves information or data snippets for a specific problem).
- A Solution Generator Agent (creates potential solutions based on the provided data).
- A Solution Evaluator Agent (validates the solutions for accuracy and relevance).

```python
grp_chat = autogen.GroupChat(
    agents=[self._data_fetcher_agent,
            self._solution_generator_agent,
            self._solution_evaluator_agent],
    messages=[],
    max_round=self.config.get("workflow.max_turns", 2),
    speaker_selection_method=custom_speaker_selection_func
)

for task in tasks_list:
    conversation_result = self._data_fetcher_agent.initiate_chat(
        self._groupchat_manager,
        message=self.message_generator,
        problem=task,
        n_results=5,
    )
    solutions.append(self.get_generated_solution(conversation_result.chat_history, "solution_generator"))
```

Currently, the agents process each task sequentially, but this approach becomes time-consuming as the number of tasks increases. I want to parallelize the execution so that agents can work on multiple tasks simultaneously.

Does the autogen.GroupChatManager support parallel execution natively? If not, are there any recommended patterns or libraries for implementing parallelism in this context?

I’d love to hear your thoughts and suggestions!

Thanks!

**Comment by ekzhu** [comment, marked as answer]
> Does the autogen.GroupChatManager support parallel execution natively? If not, are there any recommended patterns or libraries for implementing parallelism in this context?

The v0.2 group chat does not support parallel speakers. You would need to upgrade to v0.4 and use the Core API. Though at this point you need to do a bit more work to set this up. You can follow the [example patterns](https://microsoft.github.io/autogen/stable/user-guide/core-user-guide/design-patterns/intro.html), especially Mixture of Agents, for how to create parallel agents. 

The Group Chat pattern is by design a sequential execution pattern. For parallel execution I think mixture of agents is a good starting point.


UPDATE: you can now use `GraphFlow` API in AgentChat and it supports parallel flows. See https://microsoft.github.io/autogen/stable/user-guide/agentchat-user-guide/graph-flow.html

**Comment by Krishnamugundh** [comment]
hi @ekzhu , your `example patterns` link is not available. Please repost it.

**Comment by Sachin13p** [comment]
Link is still not working. Please repost @ekzhu. Thanks.

**Reply by ekzhu** [reply]
https://microsoft.github.io/autogen/stable/user-guide/core-user-guide/design-patterns/intro.html

**Reply by ekzhu** [reply]
Also updated the original answer

**Comment by rs545837** [comment]
For the parallel execution case where multiple agents modify code files simultaneously, the merge step is where things break. git's line-level merge conflicts on any file touched by two agents, even if they edited different functions.

[weave](https://github.com/ataraxy-labs/weave) is a semantic merge driver that merges at the function/class level using tree-sitter. It registers as a standard git merge driver via `.gitattributes`, so when parallel agent branches merge back, two agents adding different functions to the same file produces a clean merge instead of a conflict. On benchmarks: 31/31 clean merges vs git's 15/31.

Also ships as an MCP server with tools for coordination: `weave_claim_entity` for advisory locks on specific functions, `weave_potential_conflicts` for detecting overlapping changes before merge, and `weave_who_is_editing` to check what other agents are touching.

`brew install ataraxy-labs/tap/weave`
